registry/store/external_registry.go

```diff
@@ -27,15 +27,24 @@ type ExternalRegistry interface {
 	ListServiceInstancesByName(namespace auth.Namespace, name string) (map[string]*ServiceInstance, error)
 	ListAllServiceInstances(namespace auth.Namespace) (map[string]ServiceInstanceMap, error)
 	InsertServiceInstance(namespace auth.Namespace, instance *ServiceInstance) error
-	DeleteServiceInstance(key string) (int, error)
+	DeleteServiceInstance(namespace auth.Namespace, instanceID string) (int, error)
 }
 
+const (
+	registryRecordKey = "reg"
+	namespaceKey      = "ns"
+	instanceKey       = "inst"
+	keySeparator      = ":"
+)
+
 // DBKey represents the service instance key
 type DBKey struct {
 	Namespace  string
 	InstanceID string
 }
 
+// Database key format:
+// reg:inst:<instance id>:ns:<namespace>
 func (dbk *DBKey) String() string {
-	return fmt.Sprintf("%s%s%s", dbk.InstanceID, keySeparator, dbk.Namespace)
+	return fmt.Sprintf("%s%s%s%s%s%s%s%s%s", registryRecordKey, keySeparator, instanceKey, keySeparator, dbk.InstanceID, keySeparator, namespaceKey, keySeparator, dbk.Namespace)
```

Comment:
Assuming we'll have to support existing keys - adding/removing/moving fields in the key will be impossible, and at minimum not trivial at all. That also means that we must have control over the JSON marshaling - whitespace, fields ordering. `json.Marshal()` doesn't specify that, so we'll have to implement `MarshalJSON()` ourselves, including escaping `"` characters from the values etc. That also means larger keys (JSON overhead) and extra work to encode. The only benefit is we won't be needing to implement decoding on our own - the off-the-shelf `json.Unmarshal()` will do fine - but given the other issues, I suggest we stick to a simple (non-JSON) encoding scheme, and escape the `:` separators if present.